test-tools/src/test/java/org/terracotta/utilities/test/matchers/EventuallyTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2020 Terracotta, Inc., a Software AG company.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terracotta.utilities.test.matchers;
+
+import org.hamcrest.Matcher;
+import org.junit.Test;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.any;
+import static org.hamcrest.Matchers.both;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.stringContainsInOrder;
+import static org.junit.internal.matchers.ThrowableMessageMatcher.hasMessage;
+import static org.terracotta.utilities.test.matchers.Eventually.within;
+import static org.terracotta.utilities.test.matchers.ThrowsMatcher.threw;
+
+public class EventuallyTest {
+
+  @Test
+  public void testImmediateTrue() {
+    assertThat(() -> true, within(Duration.ZERO).is(true));
+  }
+
+  @Test
+  public void testImmediateFalse() {
+    assertThat(() -> assertThat(() -> false, within(Duration.ZERO).is(true)),
+      threw(both(any(AssertionError.class)).and(hasMessage(stringContainsInOrder(
+              "Expected:",  "<true> (within <PT0S>)", "but:", "was <false>")))));
+  }
+
+  @Test
+  public void testConditionGetsMet() {
+    long soon = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(10);
+    assertThat(System::nanoTime, within(Duration.ofSeconds(10)).matches(greaterThan(soon)));
```

Comment:
=> `assertThat(System::nanoTime, within(Duration.ofSeconds(10), greaterThan(soon)));` ?

Reply:
There's precedent for the pattern above in the Hamcrest `CombinableMatcher` classes:
```java
CombinableMatcher.CombinableBothMatcher<Boolean> both = both(is(true));
CombinableMatcher<Boolean> andMatcher = both.and(is(false));
```
I really don't like the idea of an 'infinite' timeout that leans on the `@Test` timeouts since there is no guarantee that:
a) a test timeout has been set
b) the test will respond properly to the interrupt
Also the pattern above means I can't reuse the timeout across multiple assertions (without having more redundant API).
```java
Timeout eventually = Eventually.within(Duration.ofSecond(5));
assertThat(System::nanoTime, eventually.is(5));
assertThat(System::nanoTime, eventually.matches(greaterThan(soon)));
```

Reply:
ok! agree... and the reuse of the timeout is useful indeed.